pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,998 @@
+			origin: who.clone(),
+			assets,
+			amount: amount_in,
+			amount_out,
+			discount: false,
+			discount_amount: 0_u128,
+		})
+	}
+
+	#[transactional]
+	fn execute_trade(transfer: &AMMTransfer<T::AccountId, AssetPair, Balance>) -> DispatchResult {
+		let pool_id = Self::get_pair_id(transfer.assets);
+
+		T::MultiCurrency::transfer(transfer.assets.asset_in, &transfer.origin, &pool_id, transfer.amount)?;
+		T::MultiCurrency::transfer(
+			transfer.assets.asset_out,
+			&pool_id,
+			&transfer.origin,
+			transfer.amount_out,
+		)?;
+
+		Ok(())
+	}
+
+	fn calculate_fees(amount: BalanceOf<T>) -> Result<BalanceOf<T>, DispatchError> {
+		Ok(amount
+			.just_fee(T::ExchangeFee::get())
+			.ok_or::<Error<T>>(Error::<T>::FeeAmountInvalid)?)
+	}
+
+	/// WARN: unsafe function - make sure asset_a is in pool. This function return weight in order based on asset_a id
+	fn get_weights_in_order(pool_data: &Pool<T::BlockNumber>, asset_a: AssetId) -> (LBPWeight, LBPWeight) {
+		if pool_data.last_weights.0 .0 == asset_a {
+			(pool_data.last_weights.0 .1, pool_data.last_weights.1 .1)
+		} else {
+			(pool_data.last_weights.1 .1, pool_data.last_weights.0 .1)
+		}
+	}
+}
+
+pub trait AssetPairPoolIdFor<AssetId: Sized, PoolId: Sized> {
+	fn from_assets(asset_a: AssetId, asset_b: AssetId) -> PoolId;
+}
+
+pub struct AssetPairPoolId<T: Config>(PhantomData<T>);
+
+impl<T: Config> AssetPairPoolIdFor<AssetId, PoolId<T>> for AssetPairPoolId<T>
+where
+	PoolId<T>: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+{
+	fn from_assets(asset_a: AssetId, asset_b: AssetId) -> PoolId<T> {
+		let mut buf: Vec<u8> = b"lbp".to_vec();
+
+		if asset_a < asset_b {
+			buf.extend_from_slice(&asset_a.to_le_bytes());
+			buf.extend_from_slice(&asset_b.to_le_bytes());
+		} else {
+			buf.extend_from_slice(&asset_b.to_le_bytes());
+			buf.extend_from_slice(&asset_a.to_le_bytes());
+		}
+		PoolId::<T>::unchecked_from(T::Hashing::hash(&buf[..]))
+	}
+}
+
+impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, BalanceOf<T>> for Pallet<T> {
+	fn exists(assets: AssetPair) -> bool {
+		let pair_account = T::AssetPairPoolId::from_assets(assets.asset_in, assets.asset_out);
+		<PoolAssets<T>>::contains_key(&pair_account)
+	}
+
+	fn get_pair_id(assets: AssetPair) -> T::AccountId {
+		T::AssetPairPoolId::from_assets(assets.asset_in, assets.asset_out)
+	}
+
+	fn get_pool_assets(pool_account_id: &T::AccountId) -> Option<Vec<AssetId>> {
+		match <PoolAssets<T>>::contains_key(pool_account_id) {
+			true => {
+				let assets = Self::pool_assets(pool_account_id);
+				Some(vec![assets.0, assets.1])
+			}
+			false => None,
+		}
+	}
+
+	/// Calculate spot price for given assets and amount and update pool weights if necessary
+	///
+	/// Provided assets must exist in the pool. Panic if an asset does not exist in the pool.
+	///
+	/// Return 0 if calculation overflows or weights calculation overflows.
+	fn get_spot_price_unchecked(asset_a: AssetId, asset_b: AssetId, amount: BalanceOf<T>) -> BalanceOf<T> {
```

Comment:
No tests for this. 

But as mentioned in other comment - it probably needs some refactor beause it does more that the name suggests.